lib/tesla/middleware/telemetry.ex

```diff
@@ -3,30 +3,111 @@ if Code.ensure_loaded?(:telemetry) do
 
     end
 
-    :telemetry.attach("my-tesla-telemetry", [:tesla, :request], fn event, time, meta, config ->
+    :telemetry.attach("my-tesla-telemetry", [:my_client, :tesla, :request, stop], fn event, measurements, meta, config ->
       # Do something with the event
     end)
     ```
 
+    ### Options
+
+    * `:event_prefix` - a list of atoms to prefix to the telemetry event. Defaults to `[]`
+
+    ## Telemetry Events
+
+    * `[:tesla, :request, :start]` - emitted at the beginning of the request.
+      * Measurement: `%{time: System.monotonic_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :stop]` - emitted at the end of the request.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :error]` - emitted at the end of the request when there is an error.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t, reason: term}`
+
+    * `[:tesla, :request, :exception]` - emitted at the end of the request when an exception is raised.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t, exception: Exception.t, stacktrace: Exception.stacktrace}`
+      * `[:tesla, :request]` - This event is emitted for backwards compatibility only and should be considered deprecated.
+      This event can be disabled by setting `config :tesla, Tesla.Middleware.Telemetry, disable_legacy_event: true` in your config.
+
     Please check the [telemetry](https://hexdocs.pm/telemetry/) for the further usage.
     """
 
+    @disable_legacy_event Application.get_env(:tesla, Tesla.Middleware.Telemetry,
+                            disable_legacy_event: false
+                          )[:disable_legacy_event]
+
     @doc false
-    def call(env, next, _opts) do
-      {time, res} = :timer.tc(Tesla, :run, [env, next])
-      :telemetry.execute([:tesla, :request], %{request_time: time}, %{result: res})
-      res
+    def call(env, next, opts) do
+      prefix = Keyword.get(opts, :event_prefix, [])
+      start_time = System.monotonic_time()
+
+      emit_start(env, start_time, prefix)
+
+      result =
+        try do
+          Tesla.run(env, next)
+        rescue
+          e ->
+            stacktrace = System.stacktrace()
+            metadata = %{env: env, exception: e, stacktrace: stacktrace}
+
+            :telemetry.execute(
+              prefix ++ [:tesla, :request, :exception],
+              %{duration: System.monotonic_time() - start_time},
+              metadata
+            )
+
+            reraise e, stacktrace
+        end
+
+      emit_stop(result, start_time, prefix, env)
+
+      result
+    end
+
+    defp emit_start(env, start_time, prefix) do
+      :telemetry.execute(prefix ++ [:tesla, :request, :start], %{time: start_time}, %{
+        env: env
+      })
+    end
+
+    defp emit_stop(result, start_time, prefix, req_env) do
+      duration = System.monotonic_time() - start_time
```

Comment:
Why call `System.monotonic_time()` inside `emit_stop` but outside of `emit_start`? It would be easier to reason if this was consistent (we can pass `duration` to `emit_stop` as it doesn't need `start_time` value)